File: c/leetcode/701. Insert into a Binary Search Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode *insertIntoBST(struct TreeNode *root, int val)
{
  struct TreeNode *cur = root;
  struct TreeNode node;
  node.val = val;

  while (cur)
  {
    if (val > cur->val)
    {
      if (!cur->right)
      {
        cur->right = malloc(sizeof(struct TreeNode));
        *(cur->right) = node;
        return root;
      }

      cur = cur->right;
    }

    if (val < cur->val)
    {
      if (!cur->left)
      {
        cur->left = malloc(sizeof(struct TreeNode));
        *(cur->left) = node;
        return root;
      }
      cur = cur->left;
    }
  }
  return root;
}

File: go/leetcode/701. Insert into a Binary Search Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func insertIntoBST(root *TreeNode, val int) *TreeNode {
    cur := root
    
    for cur != nil {
        if val > cur.Val {
            if cur.Right == nil {
                cur.Right = new(TreeNode)
                cur.Right.Val = val
                return root
            }
            cur = cur.Right
        }
        
        if val < cur.Val {
            if cur.Left == nil {
                cur.Left = new (TreeNode)
                cur.Left.Val = val
                return root
            }
            cur = cur.Left
        }
    }
    
    return root
}


// with pointer literal and slower
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func insertIntoBST(root *TreeNode, val int) *TreeNode {
    cur := root
    
    for cur != nil {
        if val > cur.Val {
            if cur.Right == nil {
                cur.Right = &TreeNode{val, nil, nil}
                return root
            }
            cur = cur.Right
        }
        
        if val < cur.Val {
            if cur.Left == nil {
                cur.Left = &TreeNode{val, nil, nil}
                return root
            }
            cur = cur.Left
        }
    }
    
    return root
}